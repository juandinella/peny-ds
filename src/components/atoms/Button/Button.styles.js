import { css } from 'styled-components'

import layout from '../../../styles/theme/layout'

export default css`
  display: inline-flex;
  width: 100%;
  min-width: max-content;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  transition: all 0.3s ease;
  color: var(--color-neutral-00);
  padding: var(--spacing-xs) var(--spacing-md);
  border-radius: var(--radius-lg);
  border: 1px solid transparent;
  position: relative;
  font-weight: var(--f-medium);
  text-decoration: none;

  @media screen and (min-width: ${() => layout.breakpoints.xs}) {
    padding: var(--spacing-xs) var(--spacing-md);
  }

  &:active {
    outline: 0 none;
    box-shadow: none;
  }

  &:focus {
    outline: 0 none;
    box-shadow: 0px 0px 3px 1px rgba(118, 6, 255, 0.4);
  }

  &:disabled {
    cursor: not-allowed;
    background: var(--color-neutral-100);
    color: var(--color-neutral-00);

    &:hover {
      background-color: var(--color-neutral-100);
      color: var(--color-neutral-00);
    }
  }

  ${(props) =>
    props.variant === 'primary' &&
    css`
      background-color: var(--color-primary-main);

      &:hover {
        background-color: var(--color-primary-dark);
      }

      &:active {
        transform: scale(0.98);
      }
    `}

  ${(props) =>
    props.variant === 'secondary' &&
    css`
      background-color: var(--color-secondary-main);

      &:hover {
        background-color: var(--color-secondary-dark);
      }

      &:active {
        transform: scale(0.98);
      }

      &:disabled {
        background: var(--color-neutral-80);

        &:hover {
          background-color: var(--color-neutral-80);
        }
      }
    `}

  ${(props) =>
    props.variant === 'black' &&
    css`
      background-color: var(--color-neutral-120);
      color: var(--color-neutral-00);

      &:hover {
        background-color: var(--color-neutral-100);
      }

      &:active {
        transform: scale(0.98);
      }

      &:focus {
        box-shadow: 0px 0px 3px 1px rgba(36, 36, 36, 0.4);
      }

      &:disabled {
        background: var(--color-neutral-80);

        &:hover {
          background-color: var(--color-neutral-80);
        }
      }
    `}

  ${(props) =>
    props.variant === 'white' &&
    css`
      background-color: var(--color-neutral-00);
      color: var(--color-primary-dark);

      &:hover {
        background-color: var(--color-primary-light);
      }

      &:active {
        transform: scale(0.98);
      }

      &:focus {
        box-shadow: 0px 0px 3px 1px rgba(36, 36, 36, 0.4);
      }

      &:disabled {
        background: var(--color-neutral-80);

        &:hover {
          background-color: var(--color-neutral-80);
        }
      }
    `}

  ${(props) =>
    props.variant === 'tertiary' &&
    css`
      border-radius: 0;
      padding: 0;
      color: var(--color-primary-main);
      background-color: transparent;
      max-width: max-content;

      &:hover {
        background-color: transparent !important;
        color: var(--color-primary-dark);
      }

      &:active {
        color: var(--color-primary-dark);
      }

      &:disabled {
        background: transparent;
        color: var(--color-neutral-80);

        &:hover {
          color: var(--color-neutral-80);
        }
      }

      &:focus {
        box-shadow: none;
      }
    `}


  ${(props) =>
    props.variant === 'plain' &&
    css`
      justify-content: center;
      padding: 0;

      @media screen and (min-width: ${() => layout.breakpoints.xs}) {
        padding: 0;
      }

      &:active,
      &:focus {
        outline: 0 none;
        box-shadow: none;
        transform: none;
      }
    `}

  ${(props) =>
    props.isInline &&
    css`
      min-width: 168px;
      max-width: max-content;
    `}
`
