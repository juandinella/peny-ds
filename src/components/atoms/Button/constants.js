export const options = {
  /**
   * Variant of button
   * - **primary** Primary button has the primary color
   * - **secondary** Secondary button has the secondary color
   * - **tertiary** Tertiary button is like a regular link
   * - **Plain** Plain button. No style at all
   */
  variants: ['primary', 'secondary', 'tertiary', 'plain'],
  /**
   * Determines type of button
   * - **button** The button is a clickable button.
   * - **submit** The button is a submit button (submits form data)
   * - **reset** The button is a reset button (resets the form-data to its initial values)
   */
  types: ['button', 'reset', 'submit'],
}
