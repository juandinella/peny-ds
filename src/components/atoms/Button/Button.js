import React from 'react'
import PropTypes from 'prop-types'
import styles from './Button.styles'
import { options } from './constants'
import Skeleton from '../Skeleton'
import Utils from '../../../utils'

/**
 * Used to trigger an operation.
 */

const Button = ({
  id,
  variant,
  disabled,
  onClick,
  isInline,
  type,
  children,
  href,
  active,
  fullWidthMobile,
  maxWidth,
  smallMobile,
  shouldRender,
  hideMobile,
  hideTablet,
}) => {
  const tag = typeof href === 'string' ? 'a' : 'button'

  return (
    <Skeleton
      id={`${id}_skeleton`}
      show={!Utils.emptyValue(shouldRender) && !shouldRender()}
    >
      <button
        id={id}
        data-testid={id}
        css={styles}
        variant={variant}
        disabled={disabled}
        onClick={onClick || null}
        isInline={isInline}
        type={type}
        as={tag}
        href={href}
        style={{ maxWidth: maxWidth }}
        active={active}
      >
        {children}
      </button>
    </Skeleton>
  )
}

Button.displayName = 'Button'

Button.propTypes = {
  /** Button's id */
  id: PropTypes.string,
  /** Content to be included in the button */
  children: PropTypes.oneOfType([PropTypes.string, PropTypes.node]).isRequired,
  /** This Boolean attribute prevents the user from interacting with the button */
  disabled: PropTypes.bool,
  /** Click event */
  onClick: PropTypes.func,
  /** This Boolean attribute to display an inline button */
  isInline: PropTypes.bool,
  /** Types of the Button */
  type: PropTypes.oneOf(options.types),
  /** Variatons of the buttons */
  variant: PropTypes.oneOf(options.variants),
  /** Button will render as <a> if there's href */
  href: PropTypes.string,
  active: PropTypes.bool,
  /** Set maxWidth of the button */
  maxWidth: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /**
   * Function that decides if the component should be rendered.
   *
   * @return {boolean} false if component should not be rendered and show skeleton instead. Component renders otherwise.
   */
  shouldRender: PropTypes.func,
}

Button.defaultProps = {
  variant: 'primary',
  onClick: () => null,
  isInline: false,
  disabled: false,
}

export default Button
