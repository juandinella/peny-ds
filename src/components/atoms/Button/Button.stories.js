import { options } from './constants'
import Button from './Button'

import {
  getTemplate,
  getListTemplate,
  getOptionsArgTypes,
} from '../../../helpers/storybook'

const Template = getTemplate(Button)
const ListTemplate = getListTemplate(Button)

export default {
  title: 'Atoms/Button',
  component: Button,
  args: {
    children: 'This is a Button!',
  },
  argTypes: {
    variant: getOptionsArgTypes(options.variants),
    type: getOptionsArgTypes(options.types),
    children: { control: 'text' },
  },
}

export const Default = Template.bind({})

export const Variants = ListTemplate.bind({})
Variants.args = { items: options.variants.map((variant) => ({ variant })) }

export const Types = ListTemplate.bind({})
Types.args = { items: options.types.map((type) => ({ type })) }

export const Inline = Template.bind({})
Inline.args = { isInline: true }

export const Skeleton = Template.bind({})
Skeleton.args = { shouldRender: () => false }
