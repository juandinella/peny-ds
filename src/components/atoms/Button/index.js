import Button from './Button'

export default Button
export { options } from './constants'
