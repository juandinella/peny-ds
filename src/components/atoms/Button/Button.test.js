import React from 'react'

import { render, fireEvent, cleanup } from '@testing-library/react'

import Button from './Button'

import initStoryshots from '@storybook/addon-storyshots'

initStoryshots({
  storyKindRegex: /^Atoms\/Button$/,
})

const componentProps = {
  onClick: jest.fn(),
  id: 'unique_id',
  children: <span data-testid="children">{'Go!'}</span>,
  type: 'reset',
}

afterEach(cleanup)
beforeEach(() => {
  componentProps.onClick.mockClear()
})

describe('<Button />', () => {
  test('should render component correctly', () => {
    const { queryByTestId } = render(<Button {...componentProps} />)

    const button = queryByTestId(componentProps.id)

    expect(button).toBeTruthy()
  })

  test('should render skeleton', () => {
    const { queryByTestId } = render(
      <Button {...componentProps} shouldRender={() => false} />
    )

    expect(queryByTestId(`${componentProps.id}_skeleton`)).toBeTruthy()
  })

  test('should render children correctly', () => {
    const { queryByTestId } = render(<Button {...componentProps} />)

    const children = queryByTestId('children')

    expect(children).toBeTruthy()
  })

  test('should render content correctly', () => {
    const { queryByTestId } = render(<Button {...componentProps} />)

    const button = queryByTestId(componentProps.id)

    expect(button.textContent).toEqual('Go!')
  })

  test('should ignore function if disabled', () => {
    const { container, rerender } = render(
      <Button {...componentProps} disabled />
    )

    fireEvent.click(container.firstChild)

    expect(componentProps.onClick).toHaveBeenCalledTimes(0)

    rerender(<Button {...componentProps} />)

    fireEvent.click(container.firstChild)

    expect(componentProps.onClick).toHaveBeenCalledTimes(1)
  })
  test('should call function on click', () => {
    const { container } = render(<Button {...componentProps} />)

    fireEvent.click(container.firstChild)
    fireEvent.click(container.firstChild)

    expect(componentProps.onClick).toHaveBeenCalledTimes(2)
  })

  test('should set button type correctly', () => {
    const { queryByTestId, rerender } = render(
      <Button {...componentProps} type={null} />
    )

    const button = queryByTestId(componentProps.id)

    expect(button.type).toEqual('submit')

    rerender(<Button {...componentProps} type={'reset'} />)

    expect(button.type).toEqual('reset')
  })
})
