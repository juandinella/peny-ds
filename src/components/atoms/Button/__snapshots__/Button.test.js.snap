// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Atoms/Button Default 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <button
    className="sc-x3hd59-0 hlVQpS"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
  >
    This is a Button!
  </button>
</div>
`;

exports[`Storyshots Atoms/Button Inline 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <button
    className="sc-x3hd59-0 jbAIiH"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
  >
    This is a Button!
  </button>
</div>
`;

exports[`Storyshots Atoms/Button Skeleton 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <div
    className="sc-kjnynt-0 jfZhXD"
    data-testid="undefined_skeleton"
    id="undefined_skeleton"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  >
    <button
      className="sc-x3hd59-0 hlVQpS"
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "maxWidth": undefined,
        }
      }
    >
      This is a Button!
    </button>
  </div>
</div>
`;

exports[`Storyshots Atoms/Button Types 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <button
    className="sc-x3hd59-0 hlVQpS"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
    type="button"
  >
    This is a Button!
  </button>
  <button
    className="sc-x3hd59-0 hlVQpS"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
    type="reset"
  >
    This is a Button!
  </button>
  <button
    className="sc-x3hd59-0 hlVQpS"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
    type="submit"
  >
    This is a Button!
  </button>
</div>
`;

exports[`Storyshots Atoms/Button Variants 1`] = `
<div
  style={
    Object {
      "alignContent": "flex-start",
      "display": "flex",
      "flexDirection": "column",
      "flexWrap": "wrap",
      "gap": "10px 30px",
      "height": "100%",
      "justifyContent": "flex-start",
      "maxHeight": "auto",
    }
  }
>
  <button
    className="sc-x3hd59-0 hlVQpS"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
  >
    This is a Button!
  </button>
  <button
    className="sc-x3hd59-0 dgwVJc"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
  >
    This is a Button!
  </button>
  <button
    className="sc-x3hd59-0 jgvzJV"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
  >
    This is a Button!
  </button>
  <button
    className="sc-x3hd59-0 fnuysp"
    disabled={false}
    onClick={[Function]}
    style={
      Object {
        "maxWidth": undefined,
      }
    }
  >
    This is a Button!
  </button>
</div>
`;
